### ISSUE TITLE
How to reproduce the result of Figure 2 in original paper ？

### ISSUE THREAD

**Issue by HuangZuShu** [metallurk/local_gradients_smoothing#2]

**Comment by icyham** [comment]
You can try to use `to_gx` function to transform image tensor into normalized gradient tensor. 
After that you can apply `get_gradient` to the result with certain threshold in order to get the same results as in the last image of the Figure 2.

**Comment by HuangZuShu** [comment]
Hell，@metallurk. I want to reproduce the result of the paper, but I meet two questions : 
+ My code is as follows，but the result looks different from Figure 2 (d) and more like Figure 2 (b).
+ I am also confused by the code  `r[r != r] = 0` , in `get_gradient` function . Why `r != r` is not always False?

~~~python
import torch
import torch.utils.data
import torch.cuda
import torch.optim
import torch.nn
import torch.nn.functional as F
from PIL import Image
from torchvision.transforms import ToTensor,ToPILImage
######################### device ##########################
device = torch.device("cpu")
######################### lgs ##########################
def get_gx(tensor):  # (B,C,H,W)
    B, C, W, H = tensor.shape
    sobelx = torch.tensor([[1, 0, -1], [2, 0, -2], [1, 0, -1]], device=device).float()
    sobely = sobelx.T
    G_x = F.conv2d(tensor.view(-1, 1, W, H), sobelx.view((1, 1, 3, 3)), padding=1).view(B, C, W, H)
    G_y = F.conv2d(tensor.view(-1, 1, W, H), sobely.view((1, 1, 3, 3)), padding=1).view(B, C, W, H)

    G = torch.sqrt(torch.pow(G_x, 2) + torch.pow(G_y, 2)).float()

    gx = (G-G.min())/(G.max()-G.min())

    return gx

def get_gradient(tensor, block_size=15, overlap=5, threshold=0.1):
    # tensor shape : (B,C,H,W)
    stride = block_size - overlap
    padding = []
    for v in tensor.shape[-2:]:
        left = v % stride
        if left <= overlap:
            need = overlap
        else:
            need = block_size
        pad = need - left
        padding.append(int(pad / 2))
        padding.append(pad - padding[-1])

    tensor = F.pad(tensor, tuple(padding))

    B, C, W, H = tensor.shape

    patches = tensor.unfold(2, block_size, stride).unfold(3, block_size, stride)
    # torch.Size([1, 3, 30, 30, 15, 15])
    # (patches.sum([1, -2, -1]) torch.Size([1, 30, 30])
    # print(patches.shape)
    # torch.prod(torch.tensor(patches.shape)[[1, -2, -1]])) | · | denotes the cardinality of each patch
    mask = ((patches.sum([1, -2, -1]) / torch.prod(torch.tensor(patches.shape)[[1, -2, -1]])) >= threshold).float().unsqueeze(1).unsqueeze(-1).unsqueeze(-1)

    def fold(x):
        x = x * mask
        x = x.contiguous().view(B, C, -1, block_size * block_size).permute(0, 1, 3, 2)
        x = x.contiguous().view(B, C * block_size * block_size, -1)
        return F.fold(x, output_size=(H, W), kernel_size=block_size, stride=stride)

    ox = torch.ones_like(patches)
    r = fold(patches) / fold(ox)
    r = F.pad(r, tuple(-i for i in padding))

    r[r != r] = 0

    return r

def apply_gradient(img, grad, smooth_factor=2.3):
    t = 1 - grad * smooth_factor
    # λ ∗ g(x) is clipped between 0 and 1
    # https://pytorch.org/docs/stable/generated/torch.clip.html
    clipped = torch.clip(t, 0, 1)
    # https://blog.csdn.net/Real_Brilliant/article/details/85756477
    return torch.mul(img, clipped)

# tensor ： image data
def lgs(tensor, threshold=0.1, smooth_factor=2.3, block_size=15, overlap=5):
    return apply_gradient(tensor, get_gradient(get_gx(tensor), block_size=block_size, overlap=overlap, threshold=threshold), smooth_factor=smooth_factor)

img = Image.open("figure2_a.jpg")
img = ToTensor()(img)
C,H,W = img.shape
bImgs = img.reshape(1,C,H,W)
img = ToPILImage()(get_gradient(get_gx(bImgs), block_size=15, overlap=5, threshold=0.1).reshape(C, H, W))
img.show()
~~~

**Comment by icyham** [comment]
Just set the threshold a little bit higher

**Comment by icyham** [comment]
That's my code,
I also added mean(0).unsqueeze(0) part, for image to be grayscaled like in the original paper.

~~~python
img = Image.open("a.jpg")
img = ToTensor()(img)
C,H,W = img.shape
bImgs = img.reshape(1,C,H,W)
gx = get_gx(bImgs)
figure_b = ToPILImage()(gx.squeeze().mean(0).unsqueeze(0))
thresholded_gx = get_gradient(gx, block_size=30, overlap=5, threshold=0.13)
figure_d = ToPILImage()(thresholded_gx.squeeze().mean(0).unsqueeze(0))
figure_b.show()
figure_d.show()
~~~

**Comment by HuangZuShu** [comment]
@metallurk 
It doesn't work. 
You can see the figure in google driver :   https://drive.google.com/drive/folders/1uvPlCkmz6MlNNhlIFbA60k5mv1qjZDMU?usp=sharing   .
I think there is some problem about package version.  
I also `r` in 'get_gradient' function and see some 'nan' value. 
Do you know what's the meaning of this line 'r[r != r] = 0' in 'get_gradient' function

~~~
tensor([[[[   nan,    nan,    nan,  ..., 0.3274, 0.3568, 0.3797],
          [   nan,    nan,    nan,  ..., 0.1040, 0.0493, 0.3665],
          [   nan,    nan,    nan,  ..., 0.0894, 0.0270, 0.3811],
          ...,
          [   nan,    nan,    nan,  ..., 0.2472, 0.0951, 0.3399],
          [   nan,    nan,    nan,  ..., 0.2453, 0.1326, 0.3666],
          [   nan,    nan,    nan,  ..., 0.2184, 0.2412, 0.3385]],

         [[   nan,    nan,    nan,  ..., 0.5233, 0.4977, 0.5249],
          [   nan,    nan,    nan,  ..., 0.0907, 0.0558, 0.4922],
          [   nan,    nan,    nan,  ..., 0.0545, 0.0208, 0.4783],
          ...,
          [   nan,    nan,    nan,  ..., 0.2498, 0.0984, 0.5470],
          [   nan,    nan,    nan,  ..., 0.2404, 0.1279, 0.5671],
          [   nan,    nan,    nan,  ..., 0.3816, 0.4356, 0.5431]],

         [[   nan,    nan,    nan,  ..., 0.1790, 0.2419, 0.2401],
          [   nan,    nan,    nan,  ..., 0.0984, 0.1342, 0.2175],
          [   nan,    nan,    nan,  ..., 0.0496, 0.0800, 0.2308],
          ...,
          [   nan,    nan,    nan,  ..., 0.2300, 0.0494, 0.2987],
          [   nan,    nan,    nan,  ..., 0.2191, 0.0738, 0.3216],
          [   nan,    nan,    nan,  ..., 0.1864, 0.1840, 0.2815]]]])
~~~
my code as follows .
~~~
import torch
import torch.utils.data
import torch.cuda
import torch.optim
import torch.nn
import torch.nn.functional as F
from PIL import Image
from torchvision.transforms import ToTensor,ToPILImage
######################### device ##########################
device = torch.device("cpu")
######################### lgs ##########################
def get_gx(tensor):  # (B,C,H,W)
    B, C, W, H = tensor.shape
    ######################### sobel算子求图片梯度 ： https://zhuanlan.zhihu.com/p/183926639 ##########
    sobelx = torch.tensor([[1, 0, -1], [2, 0, -2], [1, 0, -1]], device=device).float()
    sobely = sobelx.T
    G_x = F.conv2d(tensor.view(-1, 1, W, H), sobelx.view((1, 1, 3, 3)), padding=1).view(B, C, W, H)
    G_y = F.conv2d(tensor.view(-1, 1, W, H), sobely.view((1, 1, 3, 3)), padding=1).view(B, C, W, H)

    G = torch.sqrt(torch.pow(G_x, 2) + torch.pow(G_y, 2)).float()

    gx = (G-G.min())/(G.max()-G.min())

    return gx


def get_gradient(tensor, block_size=15, overlap=5, threshold=0):
    stride = block_size - overlap
    padding = []
    for v in tensor.shape[-2:]:
        left = v % stride
        if left <= overlap:
            need = overlap
        else:
            need = block_size
        pad = need - left
        padding.append(int(pad / 2))
        padding.append(pad - padding[-1])

    tensor = F.pad(tensor, tuple(padding))

    B, C, W, H = tensor.shape

    patches = tensor.unfold(2, block_size, stride).unfold(3, block_size, stride)
    mask = ((patches.sum([1, -2, -1]) / torch.prod(
        torch.tensor(patches.shape)[[1, -2, -1]])) >= threshold).float().unsqueeze(1).unsqueeze(-1).unsqueeze(-1)

    def fold(x):
        x = x * mask
        x = x.contiguous().view(B, C, -1, block_size * block_size).permute(0, 1, 3, 2)
        x = x.contiguous().view(B, C * block_size * block_size, -1)
        return F.fold(x, output_size=(H, W), kernel_size=block_size, stride=stride)

    ox = torch.ones_like(patches)
    r = fold(patches) / fold(ox)
    r = F.pad(r, tuple(-i for i in padding))
    print(r)
    r[r != r] = 0
    return r


def apply_gradient(img, grad, smooth_factor=2.3):
    t = 1 - grad * smooth_factor
    # λ ∗ g(x) is clipped between 0 and 1
    # https://pytorch.org/docs/stable/generated/torch.clip.html
    clipped = torch.clip(t, 0, 1)
    # https://blog.csdn.net/Real_Brilliant/article/details/85756477
    return torch.mul(img, clipped)

# tensor ： image data
def lgs(tensor, threshold=0.1, smooth_factor=2.3, block_size=15, overlap=5):
    return apply_gradient(tensor, get_gradient(get_gx(tensor), block_size=block_size, overlap=overlap, threshold=threshold), smooth_factor=smooth_factor)

img = Image.open("figure_2_a.jpg")
img = ToTensor()(img)
C,H,W = img.shape
bImgs = img.reshape(1,C,H,W)
gx = get_gx(bImgs)
figure_b = ToPILImage()(gx.squeeze().sum(0).unsqueeze(0))
thresholded_gx = get_gradient(gx, block_size=30, overlap=5, threshold=0.13)
figure_d = ToPILImage()(thresholded_gx.squeeze().sum(0).unsqueeze(0))
figure_b.show()
figure_b.save("figure_2_b.jpg")
figure_d.show()
figure_d.save("figure_2_d.jpg")
~~~

**Comment by icyham** [comment]
r[r != r] = 0 replaces NaNs with 0, because NaN != Nan

**Comment by icyham** [comment]
Sorry, replaced sum(0) method with mean(0).

**Comment by icyham** [comment]
Also, I think, that this is working just fine. I guess that authors of the paper made the example by hand, because block size they used in the paper is stated to be 15 pixels, but on that example it is actually like 1/5 of image's width. So they just illustrated how the method works, but the real result of this method is not exactly as it is shown in the paper.

**Comment by HuangZuShu** [comment]
I have checked the solution of examples in the paper . The solution  is 299 x 299. So I don't think the example is made by hands .  I don't know why the author didn't share the code .

**Comment by HuangZuShu** [comment]
I also don't understand these code. Can't you tell me according to the paper ? Thank you very much.
~~~
ox = torch.ones_like(patches)
r = fold(patches) / fold(ox)
r = F.pad(r, tuple(-i for i in padding))
r[r != r] = 0
~~~

**Comment by icyham** [comment]
Because of how pytorch F.fold and F.unfold operations are working, after folding the picture with overlapping regions back together, these regions' values add up with each other, and that's why we need to execute the same operation on the tensor of ones with the same shape, and divide by it the result of folding, like it is in 
~~~python
ox = torch.ones_like(patches)
r = fold(patches) / fold(ox)
~~~

~~~python
r = F.pad(r, tuple(-i for i in padding))
~~~
This deletes padding added at the start of the function

~~~python
r[r != r] = 0
~~~
And this deletes NaN values, that could have appeared at the folding dividing step

These operations are the end of getting figure 2 (d) after image was split on overlapping regions and these regions were thresholded with some value

**Comment by HuangZuShu** [comment]
@metallurk 
Thank you very much. 
I am a boy from China.
